libraries/Microsoft.Bot.Builder.Core/Middleware/StateMiddleware.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Middleware;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Middleware
+{
+    public class StateSettings
+    {
+        public bool WriteBeforeSend { get; set; } = true;
+        public bool LastWriterWins { get; set; } = true;
+    }
+
+    public abstract class StateMiddleware<StateT> : IContextCreated, ISendActivity
+        where StateT : IStoreItem, new()
+    {
+        private readonly StateSettings _settings;
+        private readonly IStorage _storage;
+
+        /// <summary>
+        /// Create statemiddleware
+        /// </summary>
+        /// <param name="name">name of the kind of state</param>
+        /// <param name="storage">storage provider to use</param>
+        /// <param name="settings">settings</param>
+        public StateMiddleware(IStorage storage, StateSettings settings = null)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+            _settings = settings ?? new StateSettings();
+        }
+
+        public abstract string GetPropertyName();
+
+        /// <summary>
+        /// The key extracted from the activity which is used to store/retrieve the state
+        /// </summary>
+        /// <param name="activity"></param>
+        /// <returns></returns>
+        public abstract string GetStorageKey(IBotContext context);
```

Comment:
>GetStorageKey [](start = 31, length = 13)

I'm doing this via a delegate passed into the constructor because I don't want to make the BotState class abstract but it's a similar concept. Are we ok with deviation here?  It makes less sense in JavaScript for everything to have to be a Class.

Reply:
delegate is good

---
In reply to: [170473579](https://github.com/Microsoft/botbuilder-dotnet/pull/193#discussion_r170473579) [](ancestors = 170473579)